Update Batch sample to use File IO

fmbenhassine opened the pull request:
This commit updates the `batch` sample to make the item reader reads data from a flat file (disk IO). This is an updated version of the Batch [getting started guide](https://spring.io/guides/gs/batch-processing/). Personally, I see no need to create two Batch samples (as mentioned in #10), but it is up to the native team to decide on this.

The goal of this PR is to port https://github.com/spring-projects-experimental/spring-native/pull/1618 from the experimental repository to this one. I took the opportunity to test the related issue (https://github.com/spring-projects-experimental/spring-native/issues/459) with the latest changes in Spring Native (as of 3e08cfd5eb116dcb1f5af819d5a2a892936f4f66) but the sample is still failing in AOT mode.

Any update on when this issue will be addressed?

sdeleuze commented:
@snicoll This is still broken with Spring Framework 6 RC1. The [original explanation](https://github.com/spring-projects-experimental/spring-native/issues/459#issue-797267585) by @mminella gives us some insights and I have been able to reproduce the issue on JVM + AOT.

It looks like this is not support by AOT transformations:
```
@StepScope
@Bean
public ItemReaderListener reader() {
   ...
}
```

This is the equivalent of:
```
@Scope(value = "step", proxyMode = ScopedProxyMode.TARGET_CLASS)
@Bean
public ItemReaderListener reader() {
   ...
}
```

Since we don't have hints for CGLIB proxies, we are kind of blocked currently.

@jhoeller Do you think we have a way to be able to detect this kind of use case AOT?

snicoll commented:
@sdeleuze let's move that to a framework issue and I'll have a look to that next week.

sdeleuze commented:
I created https://github.com/spring-projects/spring-framework/issues/29335.

sdeleuze commented:
@fmbenhassine Please test again with https://github.com/spring-projects/spring-framework/commit/6b6cf1b4a2f8b03693876dd7b534b49c48a288bd using upcoming Framework snapshots when https://ci.spring.io/teams/spring-framework/pipelines/spring-framework-6.0.x/jobs/build/builds/1316 is finished.

snicoll commented:
@fmbenhassine shouldn't the assertion be changed? This is the startup on the JVM (no AOT)

```
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::       (v3.0.0-SNAPSHOT)

2022-10-19T16:54:23.945+02:00  INFO 57918 --- [           main] com.example.batch.BatchApplication       : Starting BatchApplication using Java 17.0.4 on snicoll-a01.vmware.com with PID 57918 (/Users/snicoll/workspace/work/spring-aot-smoke-tests/batch/build/libs/batch.jar started by snicoll in /Users/snicoll/workspace/work/spring-aot-smoke-tests/batch)
2022-10-19T16:54:23.950+02:00  INFO 57918 --- [           main] com.example.batch.BatchApplication       : No active profile set, falling back to 1 default profile: "default"
2022-10-19T16:54:25.166+02:00  INFO 57918 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
2022-10-19T16:54:25.419+02:00  INFO 57918 --- [           main] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Added connection conn0: url=jdbc:h2:mem:63fbcb4b-7ac7-487e-a961-55ac29756abb user=SA
2022-10-19T16:54:25.421+02:00  INFO 57918 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Start completed.
2022-10-19T16:54:25.616+02:00  WARN 57918 --- [           main] o.s.b.c.l.AbstractListenerFactoryBean    : org.springframework.batch.item.ItemProcessor is an interface. The implementing class will not be queried for annotation based listener configurations. If using @StepScope on a @Bean method, be sure to return the implementing class so listener annotations can be used.
2022-10-19T16:54:25.740+02:00  INFO 57918 --- [           main] com.example.batch.BatchApplication       : Started BatchApplication in 2.375 seconds (process running for 2.801)
2022-10-19T16:54:25.743+02:00  INFO 57918 --- [           main] o.s.b.a.b.JobLauncherApplicationRunner   : Running default command line with: [fileName=persons.csv]
2022-10-19T16:54:25.795+02:00  INFO 57918 --- [           main] o.s.b.c.l.support.SimpleJobLauncher      : Job: [SimpleJob: [name=job]] launched with the following parameters: [{'fileName':'{value=persons.csv, type=class java.lang.String, identifying=true}'}]
2022-10-19T16:54:25.850+02:00  INFO 57918 --- [           main] o.s.batch.core.job.SimpleStepHandler     : Executing step: [step1]
2022-10-19T16:54:25.952+02:00  INFO 57918 --- [           main] o.s.batch.core.step.AbstractStep         : Step: [step1] executed in 101ms
2022-10-19T16:54:25.961+02:00  INFO 57918 --- [           main] c.e.b.JobCompletionNotificationListener  : Found <Person[firstName=JILL, lastName=DOE]> in the database.
2022-10-19T16:54:25.961+02:00  INFO 57918 --- [           main] c.e.b.JobCompletionNotificationListener  : Found <Person[firstName=JOE, lastName=DOE]> in the database.
2022-10-19T16:54:25.961+02:00  INFO 57918 --- [           main] c.e.b.JobCompletionNotificationListener  : Found <Person[firstName=JUSTIN, lastName=DOE]> in the database.
2022-10-19T16:54:25.961+02:00  INFO 57918 --- [           main] c.e.b.JobCompletionNotificationListener  : Found <Person[firstName=JANE, lastName=DOE]> in the database.
2022-10-19T16:54:25.962+02:00  INFO 57918 --- [           main] c.e.b.JobCompletionNotificationListener  : Found <Person[firstName=JOHN, lastName=DOE]> in the database.
2022-10-19T16:54:25.966+02:00  INFO 57918 --- [           main] o.s.b.c.l.support.SimpleJobLauncher      : Job: [SimpleJob: [name=job]] completed with the following parameters: [{'fileName':'{value=persons.csv, type=class java.lang.String, identifying=true}'}] and the following status: [COMPLETED] in 125ms
```

The smoke tests is looking for `Step 1 running` but I don't see it here.

sdeleuze commented:
Looks like we still have some work to support this sample.

I had to add `@RegisterReflectionForBinding(Person.class)` IMO because Batch does not infer it from`ItemProcessor<Person, Person>`. Also potentially on Batch side, `@Value("#{jobParameters['fileName']}")` require to add a `RuntimeHintsRegistrar` with `hints.reflection().registerType(StepContext.class, MemberCategory.INVOKE_PUBLIC_METHODS);`. If this is a common Spel expression, could be done by default.


More on Framework side, I had to add `hints.reflection().registerMethod(FlatFileItemReader.class.getMethod("close"), ExecutableMode.INVOKE);` since this hint is not inferred, that seems to come from `org.springframework.beans.factory.support.AbstractBeanFactory#registerDisposableBeanIfNecessary` code path.

Then after that, I get this error:
```
java.lang.NullPointerException: null
        at org.springframework.aop.framework.AdvisedSupport$MethodCacheKey.<init>(AdvisedSupport.java:578) ~[batch:6.0.0-SNAPSHOT]
        at org.springframework.aop.framework.AdvisedSupport.getInterceptorsAndDynamicInterceptionAdvice(AdvisedSupport.java:470) ~[batch:6.0.0-SNAPSHOT]
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:689) ~[na:na]
        at org.springframework.batch.item.file.FlatFileItemReader$$SpringCGLIB$$0.open(<generated>) ~[batch:na]
        at org.springframework.batch.item.support.CompositeItemStream.open(CompositeItemStream.java:124) ~[na:na]
        at org.springframework.batch.core.step.tasklet.TaskletStep.open(TaskletStep.java:293) ~[na:na]
        at org.springframework.batch.core.step.AbstractStep.execute(AbstractStep.java:217) ~[batch:5.0.0-SNAPSHOT]
        at org.springframework.batch.core.job.SimpleStepHandler.handleStep(SimpleStepHandler.java:153) ~[na:na]
        at org.springframework.batch.core.job.AbstractJob.handleStep(AbstractJob.java:408) ~[batch:5.0.0-SNAPSHOT]
        at org.springframework.batch.core.job.SimpleJob.doExecute(SimpleJob.java:132) ~[na:na]
        at org.springframework.batch.core.job.AbstractJob.execute(AbstractJob.java:307) ~[batch:5.0.0-SNAPSHOT]
        at org.springframework.batch.core.launch.support.SimpleJobLauncher$1.run(SimpleJobLauncher.java:147) ~[na:na]
        at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50) ~[na:na]
        at org.springframework.batch.core.launch.support.SimpleJobLauncher.run(SimpleJobLauncher.java:138) ~[batch:5.0.0-SNAPSHOT]
        at org.springframework.batch.core.launch.support.TaskExecutorJobLauncher.run(TaskExecutorJobLauncher.java:70) ~[na:na]
        at org.springframework.boot.autoconfigure.batch.JobLauncherApplicationRunner.execute(JobLauncherApplicationRunner.java:199) ~[batch:na]
```
So looks like more work is required on both Batch and Framework to support this sample.

jhoeller commented:
The above indicates a `null` Method coming out of the CGLIB-generated code, so likely missing reflection hints for `FlatFileItemReader.open`? It´s worth double-checking whether `GenericApplicationContext.registerProxyHintIfNecessary` is triggerered accordingly during `refreshForAotProcessing`. >This is meant to be invoked for scoped-proxy classes as well, leading to an `INVOKE_DECLARED_METHODS` hint for both the CGLIB subclass and the original user class.

sdeleuze commented:
I was able to make the sample working on native with `@RegisterReflectionForBinding(Person.class)` and 
```
hints.reflection().registerType(StepContext.class, MemberCategory.INVOKE_PUBLIC_METHODS);
hints.reflection().registerType(FlatFileItemReader.class, MemberCategory.INVOKE_PUBLIC_METHODS);
hints.resources().registerPattern("persons.csv");
hints.resources().registerPattern("schema-all.sql");
```
@jhoeller I think `GenericApplicationContext.registerProxyHintIfNecessary` is invoked since I see `"allDeclaredMethods": true` registered for `FlatFileItemReader`, but declared methods do not take in account inherited methods and `open` is defined in `AbstractItemCountingItemStreamItemReader`. So wondering if `INVOKE_DECLARED_METHODS` is right here. `INVOKE_PUBLIC_METHODS` would work here since that takes in account inherited methods, but can we stick to public method only? cc @snicoll

jhoeller commented:
@sdeleuze @snicoll `INVOKE_DECLARED_METHODS` is necessary for the proxy class itself as far as we are aware, but for the proxied user class, `INVOKE_PUBLIC_METHODS` seems appropriate indeed since a common scenario will only ever forward public method calls through the proxy - but there, public methods from all across the user class hierarchy.

fmbenhassine commented:
> I had to add @RegisterReflectionForBinding(Person.class) IMO because Batch does not infer it from ItemProcessor<Person, Person>.

This is a user class, and I'm not sure how we can infer it (with type erasure &co) and add a hint for it. I think we should not do this. Users should add hints for their domain types if necessary.

> Also potentially on Batch side, @Value("#{jobParameters['fileName']}") require to add a RuntimeHintsRegistrar with hints.reflection().registerType(StepContext.class, MemberCategory.INVOKE_PUBLIC_METHODS);. If this is a common Spel expression, could be done by default.

Yes, this is a common expression. I will add that hint as well as another one for `JobContext` by default.

snicoll commented:
> This is a user class, and I'm not sure how we can infer it (with type erasure &co) 

If Spring Batch knows that the target of an `ItemProcessor` is going to require reflection, it can easily contribute a processor that looks for beans of that type and register the necessary hints. Type erasure is not a problem since the API provides a `ResolvableType` based on the method's return type.

sdeleuze commented:
@fmbenhassine Could you please share a status update and rebase the PR on latest `main`?

fmbenhassine commented:
I added reflection hints for `StepContext` and `JobContext` as mentioned in my previous comment. I also rebased the PR on the latest main (as of b95866c1c643fd0b9684a22b4ffa5c2700f92a28).

@snicoll I will try your suggestion in another PR. I think it is OK now to have the user's hint about the custom domain type and resource. We can improve this in a later version.

The sample is working now with the latest snapshots of Batch and Framework. This is a great news! Thanks to everyone for your efforts on this!
